plugins/tpm/fu-tpm-device.c

```diff
@@ -0,0 +1,232 @@
+			     capability->data.tpmProperties.tpmProperty[0].property,
+			     query);
+		return FALSE;
+	}
+
+	if (val != NULL)
+		*val = GUINT32_FROM_BE (capability->data.tpmProperties.tpmProperty[0].value);
+	return TRUE;
+}
+
+static gchar *
+fu_plugin_tpm_get_string (ESYS_CONTEXT *ctx, guint32 query, GError **error)
+{
+	guint32 val = 0;
+	gchar result[5] = {'\0'};
+
+	/* return four bytes */
+	if (!fu_plugin_tpm_get_uint32 (ctx, query, &val, error))
+		return NULL;
+	memcpy (result, (gchar *) &val, 4);
+
+	/* convert non-ASCII into spaces */
+	for (guint i = 0; i < 4; i++) {
+		if (!g_ascii_isgraph (result[i]) && result[i] != '\0')
+			result[i] = 0x20;
+	}
+
+	return fu_common_strstrip (result);
+}
+
+static gboolean
+fu_tpm_device_setup (FuDevice *device, GError **error)
+{
+	FuTpmDevice *self = FU_TPM_DEVICE (device);
+	TSS2_RC rc;
+	guint32 tpm_type = 0;
+	guint32 version1 = 0;
+	guint32 version2 = 0;
+	g_autofree gchar *family = NULL;
+	g_autofree gchar *id1 = NULL;
+	g_autofree gchar *id2 = NULL;
+	g_autofree gchar *manufacturer = NULL;
+	g_autofree gchar *vendor1 = NULL;
+	g_autofree gchar *vendor2 = NULL;
+	g_autofree gchar *vendor3 = NULL;
+	g_autofree gchar *vendor4 = NULL;
+	g_autofree gchar *vendor_id = NULL;
+	g_autofree gchar *vendor = NULL;
+	g_autofree gchar *version = NULL;
+	g_autoptr(ESYS_CONTEXT) ctx = NULL;
+	g_autoptr(GError) error_local = NULL;
+
+	/* setup TSS */
+	rc = Esys_Initialize (&ctx, NULL, NULL);
+	if (rc != TSS2_RC_SUCCESS) {
+		g_set_error_literal (error, FWUPD_ERROR, FWUPD_ERROR_NOT_FOUND,
+				     "failed to initialize TPM library");
+		return FALSE;
+	}
+	rc = Esys_Startup (ctx, TPM2_SU_CLEAR);
+	if (rc != TSS2_RC_SUCCESS) {
+		g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
+				     "failed to initialize TPM");
+		return FALSE;
+	}
+
+	/* lookup guaranteed details from TPM */
+	family = fu_plugin_tpm_get_string (ctx, TPM2_PT_FAMILY_INDICATOR, error);
+	if (family == NULL) {
+		g_prefix_error (error, "failed to read TPM family");
+		return FALSE;
+	}
+	manufacturer = fu_plugin_tpm_get_string (ctx, TPM2_PT_MANUFACTURER, error);
+	if (manufacturer == NULL) {
+		g_prefix_error (error, "failed to read TPM manufacturer");
+		return FALSE;
+	}
+	vendor1 = fu_plugin_tpm_get_string (ctx, TPM2_PT_VENDOR_STRING_1, error);
+	if (vendor1 == NULL) {
+		g_prefix_error (error, "failed to read TPM vendor string");
+		return FALSE;
+	}
+	if (!fu_plugin_tpm_get_uint32 (ctx, TPM2_PT_VENDOR_TPM_TYPE, &tpm_type, error)) {
+		g_prefix_error (error, "failed to read TPM type");
+		return FALSE;
+	}
+
+	/* these are not guaranteed by spec and may be NULL */
+	vendor2 = fu_plugin_tpm_get_string (ctx, TPM2_PT_VENDOR_STRING_2, error);
+	vendor3 = fu_plugin_tpm_get_string (ctx, TPM2_PT_VENDOR_STRING_3, error);
+	vendor4 = fu_plugin_tpm_get_string (ctx, TPM2_PT_VENDOR_STRING_4, error);
+	vendor = g_strjoin ("", vendor1, vendor2, vendor3, vendor4, NULL);
+	fu_device_set_vendor (device, vendor);
```

Comment:
This should correspond to the manufacturer string not the vendor strings (`TPM2_PT_MANUFACTURER`).

Vendor strings should only be used for building GUID to identify the TPM.

Reply:
> Vendor strings should only be used for building GUID to identify the TPM.

So the vendor ID string is meant to be free form string that identifies the firmware stream for the device? Both examples I have here just have the TPM vendor name.

Reply:
That was my understanding of it; vendors can do whatever they want with those strings.  @flihp ,@diabonas, @AndreasFuchsSIT any other context would be helpful.

Reply:
I don't think the contents of `TPM2_PT_VENDOR_STRING` are specified by the TPM 2.0 specification - it might contain the model of the TPM (e.g. for my Infineon TPM, it reads [`SLB9670`](https://www.infineon.com/cms/de/product/security-smart-card-solutions/optiga-embedded-security-solutions/optiga-tpm/slb-9670vq2.0/)), seemingly random bytes or be completely empty. `TPM2_PT_MANUFACTURER` on the other hand should have one of the values defined in the [TCG TPM Vendor ID Registry](https://trustedcomputinggroup.org/wp-content/uploads/TCG-TPM-Vendor-ID-Registry-Version-1.01-Revision-1.00.pdf).

Reply:
> I don't think the contents of `TPM2_PT_VENDOR_STRING` are specified by the TPM 2.0 specification

Agree. The spec isn't very helpful here. It describes `TPM2_PT_VENDOR_STRING` as:

> the first four characters of the vendor ID string

which IMHO is misleading since `TPM_PT_MANUFACTURER` is also described as a `vendor ID`, though not explicitly a string:

> the vendor ID unique to each TPM manufacturer

Even then, the vendor id registry maps the hex IDs to string values: ex. `0x47 0x4F 0x4F 0x47` -> `GOOG`. The only thing to go on here is that `TPM_PT_MANUFACTURER` is guaranteed to be unique (from the vendor id registry) whereas the `TPM2_PT_VENDOR_STRING` (only one is required, up to 4) could be anything.